tensorboard/webapp/metrics/store/metrics_reducers.ts

```diff
@@ -1081,14 +1096,32 @@ const reducer = createReducer(
         state.timeSeriesData,
         linkedTimeSelection
       );
+    const nextCardStateMap = {...state.cardStateMap};
+    if (cardId) {
+      if (!nextCardStateMap[cardId]) {
+        nextCardStateMap[cardId] = {};
+      }
+      nextCardStateMap[cardId].timeSelection = linkedTimeSelection;
```

Comment:
So here's some fun NgRx behavior that I've been exploring recently: NgRx will deeply call Object.freeze() on all state after your reducer is called.

What it means for you is that, at runtime, this L1104 will silently fail to modify `nextCardStateMap[cardId].timeSelection = linkedTimeSelection;`.

You won't be able to catch this in a test because the freeze is not applied in the tests.

I recommend changing L1101-1104 to:
```
nextCardStateMap[cardId] = {
  ...nextCardStateMap[cardId],
  timeSelection: linkedTimeSelection,
};
```

It has the benefit of actually modifying timeSelection AND it handles the case where nextCardStateMap[cardId] is undefined.

Reply:
That is really interesting and makes a lot of sense. `Object.freeze` is an underutilized function.

The copy makes lots of sense and is more consistent, that is definitely how this should be done.